accepted/0000-overrides.md

```diff
@@ -0,0 +1,647 @@
+- There should only be a single copy of a given dependency in our tree, but
+  npm's dependency resolution will result in a duplicate due to version
+  conflicts.
+- There is a transitive dependency X that exhibits a bug only when used in a
+  particular way by another dependency Y.  We wish to replace that instance
+  of X, but only when included in the dependency graph of Y.
+
+## Detailed Explanation
+
+Add a field to `package.json` called `overrides`, with the following shape:
+
+```json
+{
+  "overrides": {
+    "<selection specifier>": "<result specifier or overrides set>"
+  }
+}
+```
+
+The key is a package specifier with a name.  The value is either a package
+specifier without a name, _or_ a nested overrides object.  Nested override
+objects apply to dependency resolutions within the portion of the
+dependency graph indicated by the key, and _may_ include a `"."` member
+with a string value to be applied to the node matching the override rule
+set key.
+
+### Only Root Package May Contain Overrides
+
+The `overrides` key will only be considered when it is in the root
+`package.json` file for a project.  `overrides` in installed dependencies
+will _not_ be considered in dependency tree resolution.  Thus, there is no
+cascading overrides between multiple different `package.json` files at any
+given time.
+
+Published packages may dictate their resolutions by pinning dependencies or
+using an `npm-shrinkwrap.json` file.
+
+Applying `overrides` for installed dependencies _may_ be considered in a
+future RFC.  However, there are considerable challenges in the
+implementation, user expectations, and security of such an approach.
+
+### Key Matching
+
+An override key will be considered a match if the named package specifier
+would be satisfied by the dependency being considered for resolution.
+
+For example, consider a tagged git dependency
+`project@github:user/project#v1.0.0`, which contains a `package.json` file
+with the version `1.0.0`.  In this case, the following override rule
+_would_ apply:
+
+```json
+{
+  "project@1.x": "2.0.0"
+}
+```
+
+### Overridden Dependencies Are Valid
+
+If a dependency is subject to an override that puts it outside of the
+stated dependency range for a dependent, it will still be considered valid
+if the overridden version would have satisfied the dependency.
+
+For example, consider a package `foo` that depends on `bar@1.x`, installed
+in the context of this override rule:
+
+```json
+{
+  "bar@1.x": "2.0.0"
+}
+```
+
+The resulting package tree on disk will have `foo` loading `bar@2.0.0` when
+it does `require('bar')`.  Despite this, it will _not_ be treated as an
+invalid dependency, because `bar@1.x` was valid prior to the override being
+applied.
+
+### String Overrides
+
+When the value of an override rule is a string, it is a replacement
+resolution target for resolutions matching the key.  String values _must_
+be a dependency specifier without a name.  (Aliases are supported using the
```

Comment:
It is not clear to me how are aliases treated.

For this case for instance:

```js
{
  "overrides": {
    "y@1": "1.2.3",
    "y@1.2.3": "2.3.4" // <-- this will never match anything
  }
}
```

What will this override?

Will it override this: `"y": "npm:x@1.0.0"`
or this `"x": "npm:y@1.0.0"`

Reply:
Neither of these would be overridden by this rule (at least, as the RFC is written here), because neither would match `y@1`.

It is interesting, though, and perhaps worth spelling out explicitly in the RFC, how aliases would work.

At least as I read it, if you had an override like this:

```
{
  "y@npm:x@1": "npm:x@1.2.3"
}
```

Then an installation of `y@npm:x@1.9.9` would be overridden to `y@npm:x@1.2.3`, because `y@npm:x@1.9.9` _is_ a match to `y@npm:x@1`.